Dockerfile

```diff
@@ -77,30 +84,46 @@ RUN xmlstarlet ed -L \
 # Remove unwanted files from distribution
 RUN rm -fr webapps/* *.txt *.md RELEASE-NOTES logs/ temp/ work/ bin/*.bat
 
+# hadolint ignore=DL3041
 FROM ${IMAGE_JAVA_REPO}/${IMAGE_JAVA_NAME}:${IMAGE_JAVA_TAG} AS tcnative_build-rockylinux
+ARG DISTRIB_MAJOR
 ARG JAVA_MAJOR
 ENV JAVA_HOME=/usr/lib/jvm/java-openjdk
 ARG BUILD_DIR=/build
 ARG INSTALL_DIR=/usr/local
 COPY --from=tomcat_dist /build/tcnative $BUILD_DIR/tcnative
-WORKDIR ${BUILD_DIR}/tcnative/native
+COPY --from=tomcat_dist /build/apr $BUILD_DIR/apr
 SHELL ["/bin/bash", "-euo", "pipefail", "-c"]
-RUN yum install -y gcc make openssl-devel expat-devel java-${JAVA_MAJOR}-openjdk-devel apr-devel redhat-rpm-config && yum clean all; \
-  ./configure \
+RUN yum install -y gcc make expat-devel java-${JAVA_MAJOR}-openjdk-devel redhat-rpm-config && \
+    yum clean all
+WORKDIR ${BUILD_DIR}/apr
+RUN ./configure --prefix=${INSTALL_DIR}/apr; \
+  make -j "$(nproc)"; \
+  make install
+WORKDIR ${BUILD_DIR}/tcnative/native
+RUN if [ $DISTRIB_MAJOR -eq 8 ]; then dnf install -y dnf-plugins-core; \
+    dnf config-manager -y --set-enabled powertools; \
+    dnf install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm; \
+    dnf install -y openssl3-devel; \
+    ln -s /usr/include/openssl3/openssl /usr/include/openssl; \
+    export LIBS="-L/usr/lib64/openssl3 -Wl,-rpath,/usr/lib64/openssl3 -lssl -lcrypto"; \
+    export CFLAGS="-I/usr/include/openssl3"; \
+    else dnf install -y openssl-devel; \
+    fi; \
+    dnf clean all; \
+    ./configure \
     --libdir=${INSTALL_DIR}/tcnative \
-    --with-apr=/usr/bin/apr-1-config \
-    --with-java-home="$JAVA_HOME"; \
+    --with-apr=${INSTALL_DIR}/apr/bin/apr-1-config \
+    --with-java-home="$JAVA_HOME" \
+    --disable-openssl-version-check; \
   make -j "$(nproc)"; \
   make install
 
 # hadolint ignore=DL3006
 FROM tcnative_build-${DISTRIB_NAME} AS tcnative_build
```

Comment:
> Default value for global ARG results in an empty or invalid base image name

we have plenty of linter message about this (even in bakery) - I am starting to think we should provide a sensible defaults for arguments - I guess it's because forgetting to provide a "mandatory" arg is going to fail the build with some hard to figure message - see for more https://docs.docker.com/reference/build-checks/invalid-default-arg-in-from/

Reply:
Doing it here should be fine indeed, will try it. For bakery however the default values are managed in the bake file. I'm not sure setting ARG's defaults in Dockerfile would be DRY